accord-core/src/main/java/accord/impl/progresslog/WaitingState.java

```diff
@@ -0,0 +1,660 @@
+
+            int roundSize = awaitRoundSize(slicedRoute);
+            int roundIndex = awaitRoundIndex(roundSize);
+            int updateBitSet = roundCallbackBitSet(owner, txnId, from, slicedRoute, callbackId, roundIndex, roundSize);
+            if (updateBitSet == 0)
+                return;
+
+            int bitSet = awaitBitSet(roundSize);
+            bitSet &= ~updateBitSet;
+            setAwaitBitSet(bitSet, roundSize);
+
+            if (bitSet == 0)
+                runInternal(safeStore, safeCommand, owner);
+        }
+    }
+
+    // TODO (expected): use back-off counter here
+    private void retry(SafeCommandStore safeStore, SafeCommand safeCommand, DefaultProgressLog owner, BlockedUntil blockedUntil)
+    {
+        if (!contactEveryone())
+        {
+            setContactEveryone(true);
+            // try again immediately with a query to all eligible replicas
+            runInternal(safeStore, safeCommand, owner);
+        }
+        else
+        {
+            // queue a retry
+            set(safeStore, owner, blockedUntil, Queued);
+        }
+    }
+
+    static void fetchRoute(DefaultProgressLog owner, BlockedUntil blockedUntil, TxnId txnId, Timestamp executeAt, EpochSupplier toLocalEpoch, Participants<?> fetchKeys)
+    {
+        // TODO (desired): fetch only the route
+        // we MUSt allocate before calling withEpoch to register cancellation, as async
+        BiConsumer<FetchData.FetchResult, Throwable> invoker = invokeWaitingCallback(owner, txnId, blockedUntil, WaitingState::fetchRouteCallback);
+        owner.node().withEpoch(blockedUntil.fetchEpoch(txnId, executeAt), invoker, () -> {
+            FetchData.fetch(blockedUntil.minSaveStatus.known, owner.node(), txnId, fetchKeys, toLocalEpoch, executeAt, invoker);
+        });
+    }
+
+    static void fetch(DefaultProgressLog owner, BlockedUntil blockedUntil, TxnId txnId, Timestamp executeAt, EpochSupplier toLocalEpoch, Route<?> slicedRoute, Route<?> fetchRoute)
+    {
+        Invariants.checkState(!slicedRoute.isEmpty());
+        // we MUSt allocate before calling withEpoch to register cancellation, as async
+        BiConsumer<FetchData.FetchResult, Throwable> invoker = invokeWaitingCallback(owner, txnId, blockedUntil, WaitingState::fetchCallback);
+        owner.node().withEpoch(blockedUntil.fetchEpoch(txnId, executeAt), invoker, () -> {
+            FetchData.fetchSpecific(blockedUntil.minSaveStatus.known, owner.node(), txnId, fetchRoute, slicedRoute, toLocalEpoch, executeAt, invoker);
+        });
+    }
+
+    void awaitHomeKey(DefaultProgressLog owner, BlockedUntil blockedUntil, TxnId txnId, Timestamp executeAt, Route<?> route)
+    {
+        // TODO (expected): special-case when this shard is home key to avoid remote messages
+        await(owner, blockedUntil, txnId, executeAt, route.homeKeyOnlyRoute(), Integer.MAX_VALUE, WaitingState::synchronousAwaitHomeCallback);
```

Comment:
This passes `Integer.MAX_VALUE` as the `callbackid`, is that supposed to be -1?

Reply:
Nope, it's meant to be an id we don't use for walking the shards, so we just use the largest possible integer on the assumption we will never have a `Route` with `Integer.MAX_VALUE` keys in it.